spec.html

```diff
@@ -303,17 +309,32 @@ <h1>SharedArrayBuffer.prototype.grow ( _newLength_ )</h1>
+        1. NOTE: In implementations, _rawCurrentByteLengthBytesRead_ is the result of a load-link, of a load-exclusive, or of an operand of a read-modify-write instruction on the underlying hardware. The nondeterminism is a semantic prescription of the memory model to describe observable behaviour of hardware with weak consistency.
+        1. Let _isLittleEndian_ be the value of the [[LittleEndian]] field of the surrounding agent's Agent Record.
+        1. Let _currentBytesLength_ be RawBytesToNumeric(~BigUint64~, _rawCurrentByteLengthBytesRead_, _isLittleEndian_).
+        1. If _newByteLength_ &lt; _currentByteLength_ or _newByteLength_ &gt; _O_.[[ArrayBufferMaxByteLength]], throw a *RangeError* exception.
+        1. NOTE: The above may fail due to concurrent calls to SharedArrayBuffer.prototype.grow racing. For example, if one agent is evaluating `sab.grow(10)` and another agent is evaluating `sab.grow(20)` concurrently, either may fail due to a race while atomically updating the length.
         1. If _newByteLength_ &ne; _currentByteLength_, then
-          1. Note: Resizes to the same length explicitly do nothing to avoid gratuitous synchronization.
+          1. NOTE: Resizes to the same length explicitly do nothing to avoid gratuitous synchronization.
+          1. Let _execution_ be the [[CandidateExecution]] field of the surrounding agent's Agent Record.
+          1. Let _eventList_ be the [[EventList]] field of the element in _execution_.[[EventsRecords]] whose [[AgentSignifier]] is AgentSignifier().
+          1. Let _second_ be a new read-modify-write modification function with parameters (_oldBytes_, _newBytes_) that captures nothing and performs the following steps atomically when called:
+            1. Return _newBytes_.
           1. Let _byteLengthBlock_ be _O_.[[ArrayBufferByteLengthData]].
-          1. Perform SetValueInBuffer(_byteLengthBlock_, 0, ~Uint64~, 𝔽(_newByteLength_), *true*, ~SeqCst~).
+          1. Let _newByteLengthBytes_ be NumericToRawBytes(~BigUint64~, 𝔽(_newByteLength_), _isLittleEndian_).
+          1. Let _event_ be ReadModifyWriteSharedMemory { [[Order]]: ~SeqCst~, [[NoTear]]: *true*, [[Block]]: _byteLengthBlock_, [[ByteIndex]]: 0, [[ElementSize]]: 8, [[Payload]]: _newByteLengthBytes_, [[ModifyOp]]: _second_ }.
+          1. Append _event_ to _eventList_.
+          1. Append Chosen Value Record { [[Event]]: _event_, [[ChosenValue]]: _rawCurrentByteLengthBytesRead_ } to _execution_.[[ChosenValues]].
+          1. NOTE: The new memory is already zeroed, as a _O_.[[ArrayBufferMaxByteLength]] sized Data Block is already allocated. This is a specification mechanism, an implementation is not required to reserve _O_.[[ArrayBufferMaxByteLength]] bytes of physical memory.
         1. Return *undefined*.
       </emu-alg>
+      <emu-note>
+        <p>Many of the above steps are shared with the algorithm steps of Atomics.compareExchange and should be refactored when merged into the full specification.</p>
```

Comment:
Maybe I'm still misunderstanding, since there is a compare still for if the new length is legal and hasn't been changed since it was read to avoid. Wouldn't an `exchange` not guard the OP's issue in #39, where a race can result in an accidental shrink?

(I've never manually programmed LL/SC, so I was perhaps incorrectly thinking LL/SC can actually express the CAS here where the compare isn't just a ==.)

Reply:
Ah, I think I see the source of my confusion. The test in line 319 (`If new < current then throw`) can succeed/throw entirely non-deterministically, since `_currentByteLength_` is picked non-deterministically, and if the operation actually throws in this way, no event is currently emitted recording the read value of `_currentByteLength_`. There's a similar issue with the `If new == current` test to skip the emission of the RMW.

In Wasm, growth failure is just entirely non-deterministic (even prior to shared memory), so the Wasm specification looks like

on growth
- either fail (with no distinction as to whether OOB or OOM error), or
- non-deterministically pick a current length such that current+growth <= max and emit the corresponding RMW event

Here there's some more subtlety. IIUC JS implementations are not required to throw a RangeError on alloc failure (although they are advised to). Presumably you only want the spec to _mandate_  throwing a RangeError if `new < current`. In any case, you also want to only be idempotent if `new != current`.

I think the simplest solution is to make sure that this operation always emits at least an SC read event if it fails either due to the `new < current` check, or skips due to the `new = current` check. It's still permitted (I guess) for the operation to emit nothing if there's an OOM error (which isn't specified explicitly).

TL;DR The current alg does not emit a read event on OOB failure, meaning it's more non-deterministic than it should be. If the alg were to emit a read event on failure, a parallel could be drawn with `Atomics.compareExchange`.

Reply:
I see, thanks for spelling out the thinking process here, very helpful.

Emitting an SC read on failure sounds fine to me.

As for are OOMs mandated to throw, my intention was yes but this was lost for the SAB case because of the spec fiction that SABs allocate a _maxLength_ shared data block up front. ABs have this point because the spec mechanism is creating a new data block and copying over the old contents. I wonder what the best thing to do here is to convey that OOM errors must throw...